keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+  stage: Drain
+  drainPlan:
+    - podPriority: 5000
+      podType: Default
+    - podPriority: 15000
+      podType: Default
+    - podPriority: 3000
+      podType: DaemonSet
+  ...
+status:
+  drainStatus:
+    podsPendingEvacuation: 91
+    podsEvacuating: 39
+    drainMessage: "Evacuating"
+    reachedDrainTargets:
+      - podPriority: 10000
+        podType: Default
+  ...
+---
+apiVersion: v1alpha1
+kind: NodeMaintenance
+metadata:
+  name: "maintenance-b"
+spec:
+  nodeSelector:
+    # selects nodes one and three
+  stage: Drain
+  drainPlan:
+    - podPriority: 10000
+      podType: Default
+    - podPriority: 15000
+      podType: Default
+    - podPriority: 4000
+      podType: DaemonSet
+  ...
+status:
+  drainStatus:
+    podsPendingEvacuation: 115
+    podsEvacuating: 30
+    drainMessage: "Evacuating"
+    reachedDrainTargets:
+      - podPriority: 10000
+        podType: Default
+  ...
+---
+apiVersion: v1
+kind: Node
+metadata:
+  name: "one"
+status:
+  maintenanceStatus:
+    podsPendingEvacuation: 70
+    podsEvacuating: 30
+    drainMessage: "Evacuating (limited by maintenance-b)"
+    drainTargets:
+      - podPriority: 10000
+        podType: Default
+  ...
+---
+apiVersion: v1
+kind: Node
+metadata:
+  name: "two"
+status:
+  maintenanceStatus:
+    podsPendingEvacuation: 21
+    podsEvacuating: 9
+    drainMessage: "Evacuating"
+    drainTargets:
+      - podPriority: 15000
+        podType: Default
+  ...
+---
+apiVersion: v1
+kind: Node
+metadata:
+  name: "three"
+status:
+  maintenanceStatus:
+    podsPendingEvacuation: 45
+    podsEvacuating: 0
+    drainMessage: "Waiting for maintenance-b."
+    drainTargets:
+      - podPriority: 10000
+        podType: Default
+  ...
+```
+
+The progress of the drain should not be backtracked. If an intersecting `maintenance-c` is created,
+the node one progress should stay the same regardless of the node maintenance drainPlan.
+
+
+```yaml
+apiVersion: v1alpha1
+kind: NodeMaintenance
+metadata:
+  name: "maintenance-c"
+spec:
+  nodeSelector:
+    # selects nodes one and four
+  stage: Drain
+  drainPlan:
+    - podPriority: 2000
+      podType: Default
+    - podPriority: 15000
+      podType: Default
+  ...
+status:
+  drainStatus:
+    podsPendingEvacuation: 90
+    podsEvacuating: 35
+    drainMessage: "Evacuating"
+    reachedDrainTargets:
+      - podPriority: 2000
+        podType: Default
+  ...
+---
+apiVersion: v1
+kind: Node
+metadata:
+  name: "one"
+status:
+  maintenanceStatus:
+    podsPendingEvacuation: 70
+    podsEvacuating: 30
+    drainMessage: "Evacuating (limited by maintenance-b, maintenance-c)"
+    drainTargets:
+      - podPriority: 10000
+        podType: Default
+  ...
+---
+apiVersion: v1
+kind: Node
+metadata:
+  name: "four"
+status:
+  maintenanceStatus:
+    podsPendingEvacuation: 20
+    podsEvacuating: 5
+    drainMessage: "Evacuating"
+    drainTargets:
+      - podPriority: 2000
+        podType: Default
+  ...
+---
+```
```

Comment:
+1 
It could be really help having an example that shows how a user can figure out why a node drain is blocked when multiple node maintenance apply.
(also considering all the moving parts 😅, or at least Evacuation, Eviction, DaemonSet controller, kubelet)

Reply:
Each Node will have its status and the NodeMaintenace should be the aggregate. Please take into account that these are not final designs and may change considerably from the final API.  There is still no consensus if there should be m:n mapping, for example.

We can add the debugging parts once we move into the PRR phase.